code/modules/mob/living/living_inventory.dm

```diff
@@ -0,0 +1,99 @@
+/mob/living/set_inventory_slots(var/list/new_slots, var/preserve_hands)
+
+	var/list/old_slots = inventory_slots
+	inventory_slots = null
+
+	// Keep hands if desired.
+	if(preserve_hands && LAZYLEN(old_slots))
+		for(var/slot in get_held_item_slots())
+			if(slot in old_slots)
+				LAZYSET(inventory_slots, slot, old_slots[slot])
+
+	// Check if we need to replace any existing slots.
+	for(var/new_slot_id in new_slots)
+		var/datum/inventory_slot/new_slot = LAZYACCESS(new_slots, new_slot_id)
+		var/datum/inventory_slot/old_slot = LAZYACCESS(old_slots, new_slot_id)
+		if(old_slot)
+			new_slot.set_slot(old_slot.get_equipped_item())
+			qdel(old_slot)
+		LAZYSET(inventory_slots, new_slot.slot_id, new_slot)
+
+/mob/living/get_held_item_slots()
+	return held_item_slots
+
+/mob/living/clear_inventory_slots(var/ignore_hands)
+	for(var/slot in inventory_slots)
+		if(!ignore_hands || !(slot in held_item_slots))
+			remove_inventory_slot(slot)
+	QDEL_NULL_LIST(inventory_slots)
+
+/mob/living/add_inventory_slot(var/datum/inventory_slot/inv_slot)
+	. = ..()
+	if(.)
+		LAZYSET(inventory_slots, inv_slot.slot_id, inv_slot)
+
+/mob/living/remove_inventory_slot(var/slot_id)
+	var/datum/inventory_slot/inv_slot = LAZYACCESS(inventory_slots, slot_id)
+	if(inv_slot)
+		var/held = inv_slot.get_equipped_item()
+		if(held)
+			drop_from_inventory(held)
+		qdel(inv_slot)
+	LAZYREMOVE(inventory_slots, slot_id)
```

Comment:
Does this ever remove it from `held_item_slots`? That feels like it'd cause hand slots to have a hanging reference that makes them fail GC.